### PR TITLE
Migrate (legacy) workflow command to click

### LINKED ISSUE #1597: verdi click: migrate verdi workflow
Migrate the verdi command `workflow` to use the new `click` framework

### LINKED ISSUE #1649: add 'deprecated' decorator for click
As we are going to deprecate several commands for aiida 1.0, we need a way to specify semantically that a command is deprecated.

Note: Using  `echo.echo_deprecated` is not enough, since we want e.g. those commands not to show up in autocompletion.

### LINKED ISSUE #1671: Legacy Workflows: sqla implementation's get_report fails if report does not exist
## Expected behaviour
`get_report()` should return a list of strings or empty list

```
from aiida.workflows.test import WFSimpleWithSubWf
workflow = WFSimpleWithSubWf()
workflow.store()
workflow.start()
report = workflow.get_report()  # should return a list of strings or empty list if there are no reports
```

## Observed behaviour
`get_report()` fails in the following case

### Test
Shortened for the sake of legibility:

```
class TestVerdiLegacyWorkflow(AiidaTestCase):

    @classmethod
    def setUpClass(cls):
        from aiida.workflows.test import WFTestEmpty, WFTestSimpleWithSubWF

        super(TestVerdiLegacyWorkflow, cls).setUpClass()
        cls.runner = CliRunner()

        cls.super_workflow = WFTestSimpleWithSubWF()
        cls.super_workflow.store()
        cls.super_workflow.start()

    def test_workflow_report(self):
        result = self.runner.invoke(workflow_report, [str(self.super_workflow.uuid)])
        self.assertIsNone(result.exception, msg=debug_msg(result))
        self.assertIn(str(self.super_workflow.pk), result.output)

```

### Output

```
======================================================================
FAIL: test_workflow_report (aiida.backends.tests.cmdline.commands.test_workflow.TestVerdiLegacyWorkflow)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ricohaeuselmann/Code/aiida-fork/aiida/backends/tests/cmdline/commands/test_workflow.py", line 68, in test_workflow_report
    self.assertIsNone(result.exception, msg=debug_msg(result))
AssertionError: Traceback (most recent call last):
  File "/Users/ricohaeuselmann/.venvwrap/aiida-dev-1/lib/python2.7/site-packages/click/testing.py", line 279, in invoke
    prog_name=self.get_default_prog_name(cli), **extra)
  File "/Users/ricohaeuselmann/.venvwrap/aiida-dev-1/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/ricohaeuselmann/.venvwrap/aiida-dev-1/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ricohaeuselmann/.venvwrap/aiida-dev-1/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/ricohaeuselmann/Code/aiida-fork/aiida/cmdline/utils/decorators.py", line 183, in decorated_function
    return function(*args, **kwargs)
  File "/Users/ricohaeuselmann/Code/aiida-fork/aiida/cmdline/commands/workflow.py", line 130, in workflow_report
    echo.echo('\n'.join(workflow.get_report()))
  File "/Users/ricohaeuselmann/Code/aiida-fork/aiida/orm/implementation/sqlalchemy/workflow.py", line 499, in get_report
    return self.dbworkflowinstance.report.splitlines()
AttributeError: 'NoneType' object has no attribute 'splitlines'
```

### PR DESCRIPTION
**Pull request by DropD**
fix #1597 
fix #1649 
fix #1671

### CODE REVIEW

File: aiida/cmdline/commands/workflow.py [hunk 1]
```diff
@@ -28,214 +37,131 @@ def __init__(self):
...
+            if state not in ['FINISHED', 'ERROR']:
+                workflows.extend(Workflow_.query(state=state))
+    return [workflow for workflow in workflows]
+
+
+@verdi_workflow.command('logshow')
+@click.argument('workflows', type=LegacyWorkflowParamType(), nargs=-1)
+@deprecated_command(DEPRECATION_MSG)
+def workflow_logshow(workflows):
+    """Show the for each in a list of WORKFLOWS."""
```
Comment (sphuber) [inline aiida/cmdline/commands/workflow.py]: Docstring sentence does not read well

File: aiida/cmdline/commands/workflow.py [hunk 2]
```diff
@@ -28,214 +37,131 @@ def __init__(self):
...
+        state = workflow.get_state()
+        echo.echo('*** {pk}{label}: {state}'.format(pk=workflow.pk, label=label_str, state=state))
+
+        if workflow.get_report():
+            echo.echo('Print the report with `verdi workflow report {}`'.format(workflow.pk))
+        else:
+            echo.echo('*** Report is empty')
+
+        if log_messages:
+            print '*** {} LOG MESSAGES:'.format(len(log_messages))
```
Comment (sphuber) [inline aiida/cmdline/commands/workflow.py]: swap `print` for `echo.echo`

File: aiida/cmdline/utils/decorators.py [hunk 3]
```diff
@@ -156,4 +156,30 @@ def decorated_function(*args, **kwargs):
...
+
+
+def deprecated_command(message):
+    def decorator(function):
+        """
+        Function decorator that echoes a deprecation warning before doing anything else in a click commad.
+
+        Example::
+
+            @click.commad()
```
Comment (sphuber) [inline aiida/cmdline/utils/decorators.py]: typo

**Comment by sphuber** [comment]
Only minor nitpicks, but since it is deprecated command don't want to force you to fix those. If you have to add another commit anyway, to bring branch up to date, you can include these

**Comment by giovannipizzi** [comment]
@szoupanos should have more experience with SQLA and old workflows. @szoupanos do your tests show such a  behaviour? Could you see if it is easy to fix? Probably we should just make sure that `self.dbworkflowinstance.report` returns `""` rather than `None`?

**Comment by DropD** [comment]
@giovannipizzi @szoupanos I created #1671 for this and fixed it in an additional commit.

File: aiida/orm/implementation/sqlalchemy/workflow.py [hunk 4]
```diff
@@ -496,7 +496,8 @@ def get_report(self):
           This is not the case anymore.
         :return: a list of strings
         """
-        return self.dbworkflowinstance.report.splitlines()
```
Comment (DropD) [inline aiida/orm/implementation/sqlalchemy/workflow.py]: This was the culprit leading to #1671

**Comment by szoupanos** [comment]
@DropD if I understand well the error is fixed under SQLA?
@giovannipizzi (and Rico too), the old workflows were tested under SQLA and worked (before passing to the new daemon) but maybe this command was not used. The tests don't run automatically (it may need a bit of time to start them now - remember where I stand) and maybe it makes more sense to try to add them directly to the CI that @giovannipizzi setup (and run them from there).

**Pull request merged**